client/lib/route/path.js

```diff
@@ -63,12 +63,19 @@ function addSiteFragment( path, site ) {
 	return pieces.join( '/' );
 }
 
-function sectionify( path ) {
+function sectionify( path, siteFragment ) {
 	let basePath = path.split( '?' )[ 0 ];
-	const site = getSiteFragment( basePath );
 
-	if ( site ) {
-		basePath = trailingslashit( basePath ).replace( '/' + site + '/', '/' );
+	// Sometimes the caller knows better than `getSiteFragment` what the `siteFragment` is.
+	// For example, when the `:site` parameter is not the last or second-last part of the route
+	// and is retrieved from `context.params.site`. In that case, it can pass the `siteFragment`
+	// explicitly as the second parameter. We call `getSiteFragment` only as a fallback.
+	if ( ! siteFragment ) {
+		siteFragment = getSiteFragment( basePath );
```

Comment:
Not a blocker, but there's already a `route.getSiteFragment()` fallback when you set `siteFragment` inside `siteSelection`. Should we change the other calls to `sectionify()` in `/my-sites/controller.js` to use this as well so we can remove the fallback here?

Or... would it be cleaner to just modify `sectionify()` to accept the full context object and do the fallback here? That'd require a lot of changes because I see `sectionify()` called 90 times in the code. Or last, maybe check to see if `path` is an object or string, and treat is as the context object if it's an object.

None of these are blockers though, this is probably the simplest fix.